File: examples/test_iterator.py
"""Test iterator protocol for v1.3.0"""

class Range:
    """A simple range iterator"""
    
    def __init__(self, start: int, stop: int):
        self.start = start
        self.stop = stop
        self.current = start
    
    def __iter__(self):
        """Return self as iterator"""
        return self
    
    def __next__(self) -> int:
        """Get next value"""
        if self.current < self.stop:
            value = self.current
            self.current += 1
            return value
        else:
            # In Python, would raise StopIteration
            return -1  # Simplified for now

def test_custom_iterator():
    """Test custom iterator"""
    r = Range(0, 5)
    total = 0
    
    # Manual iteration for now
    value = r.__next__()
    while value != -1:
        total += value
        value = r.__next__()
    
    return total

def test_for_with_iterator():
    """Test for loop with iterator"""
    r = Range(0, 5)
    total = 0
    
    # This should work once we have proper iterator support
    for i in r:
        total += i
    
    return total

File: examples/test_iterator.rs
#[derive(Debug, Clone)]
pub struct Range {
    pub start: i32,
    pub stop: i32,
    pub current: i32,
}
impl Range {
    pub fn new(start: i32, stop: i32) -> Self {
        Self {
            start,
            stop,
            current: 0,
        }
    }
    pub fn __iter__(&self) -> &Self {
        return self;
    }
    pub fn __next__(&mut self) -> i32 {
        if self.current < self.stop {
            let mut value = self.current;
            self.current = self.current + 1;
            return value;
        } else {
            return -1;
        };
    }
}
#[doc = "Test custom iterator"]
#[doc = " Depyler: verified panic-free"]
pub fn test_custom_iterator() {
    let mut r = Range::new(0, 5);
    let mut total = 0;
    let mut value = r.__next__();
    while value != -1 {
        total = total + value;
        value = r.__next__();
    }
    total
}
#[doc = "Test for loop with iterator"]
#[doc = " Depyler: verified panic-free"]
pub fn test_for_with_iterator() {
    let mut r = Range::new(0, 5);
    let mut total = 0;
    for i in r.iter().cloned() {
        total = total + i;
    }
    total
}
